Feature/add ticket system

FabioGaming opened the pull request:
Finalized the ticket system and added some additional documentation on the README file, ticket system should now be fully operational but future updates are required.

FabioGaming committed:
added ticket base

FabioGaming committed:
Did more stuff for tickets

FabioGaming committed:
added function base for ticket alias adding

FabioGaming committed:
added command base for ticket initiator message & added new keys to DBManager

[ + ] Added new Keys to DBManager on TicketCategories so they can be used more fluently
[ + ] Added command base for mod-inittickets

FabioGaming committed:
adapted Ticket related stuff to work with Dapper mapping

[ # ] Adapted Object mapping to work with Dapper

FabioGaming committed:
removed using statements, fixed typos and added more SQLite Handling

[ + ] Made some Columns UNIQUE
[ # ] Dapper returns now use Parameters for safe queries
[ # ] Task<int> has been changed to int
[ + ] Added AddOrUpdateTicketCategory(string, string, string)
[ + ] Added SetTicektAutoResponse(string, string)
[ + ] Added SetTicketDataDescription(string, string)
[ - ] Removed unused namepsaces from using list

FabioGaming committed:
Added more DBManager & Command Bases

[ + ] Added DBManager.SetTicketTopicText(string, string)
[ + ] Added Commands.AddOrUpdateTicketCategory(string, string, string)
[ # ] Autofills.TicketAutoCompleteHandler.TicketAutoCompleteLoader now uses TicketDataName as value
[ - ] Removed Utils.DefaultSloganText (and adapted commands)

FabioGaming committed:
Added /init-ticket command base & refactored some code

[ - ] Removed Commands.AddTicketAlias() (/mod-addticketalias) as it will not be used as of now
[ # ] Finalized Commands.InitTicketMessage() (/mod-inittickets)
[ + ] Added base for Components.HandleTicketMenu()
[ - ] Removed unused using statements

FabioGaming committed:
Added more base code

[ - ] Removed unused using statements
[ # ] HandleTicketMenu is almost ready
[ + ] Added HandleTicketDismiss to handle auto response deletions
[ + ] Added HandleTicketForce base function to handle forced tickets

FabioGaming committed:
continued Ticket System

[ + ] Added Commands.ModifyRoleRule (/mod-setrolerule) to add another layer of dynamic config
[ # ] Fixed Commands.AddOrUpdateTicketCategory (/mod-settickettopic), it will now respond properly
[ + ] Added Commands.SetTicketAutoResponse (/mod-setticketautoresponse)
[ + ] Added DBManager support for new Role system
[ # ] DBManager.UpdateInternalConfig() now includes Role config changes
[ + ] Added DBManager.ModifyRoleConfig()
[ # ] UserJoinHandler no longer uses old template string
[ + ] Added Responses.TicketInitMessage(string, string, ulong)
[ + ] Added Utils.CreateTicketChannel(InteactionModuleBase) to properly initialize ticket channels

FabioGaming committed:
Finalized Ticket System & Formatted Code

[ # ] Formatted all files
[ # ] FixedTicketCategoryAutoCompleteHandler so it will no longer return invalid topic codes
[ + ] Added all Ticket options to: open, close and force tickets
[ # ] Fixed DBManager since it had some typos which created invalid SQL Queries
[ + ] Added TicketCreationModalModel as modal class for the ticket Modal
[ # ] Fixed invalid Variable name in Objects.TicketCategory, since it was not found by Dapper Queries
[ + ] Added Identifier Wiki section in README.md for easy access to interrnal values that are required for the bot to work

FabioGaming reviewed (commented):
LGTM;

lenajenichen reviewed (changes requested):
See GitHub Action comments. Change them first before approval.

FabioGaming committed:
adjusted some variables

[ # ] Commands.EchoMessage and Commands.EchoMessageFromFile now use nullable values to not trigger warnings
[ # ] Program.OnLog has been moved to DEBUG only, some variables have also been marked nullable
[ # ] Utils.Config is now nullable

lenajenichen commented on LuminateDiscordBot/Commands.cs:
```diff
@@ -24,23 +17,41 @@ public async Task ModifyChannelRules([Summary("channel_identifier", "The interna
+            embed.Color = Color.Blue;
+            embed.Footer = new EmbedFooterBuilder()
+            {
+                Text = $"Luminate - Your ideas shine bright"
```
String exists 2 times in the code already. Maybe a variable would be usefull for future usage.

lenajenichen commented on LuminateDiscordBot/Commands.cs:
```diff
@@ -24,23 +17,41 @@ public async Task ModifyChannelRules([Summary("channel_identifier", "The interna
             embed.Footer = new EmbedFooterBuilder()
             {
-                Text = $"Luminate - {Utils.DefaultSloganText}"
+                Text = $"Luminate - Your ideas shine bright"
```
Same here the string should be changed

lenajenichen commented on LuminateDiscordBot/Commands.cs:
```diff
@@ -52,7 +63,7 @@ public async Task EchoMessage(string message, bool asEmbed = false, string embed
                 echo.Footer = new EmbedFooterBuilder()
                 {
-                    Text = $"Luminate - {Utils.DefaultSloganText}"
+                    Text = $"Luminate - Your ideas shine bright"
```
Same here the string should be changed

lenajenichen commented on LuminateDiscordBot/Commands.cs:
```diff
@@ -63,10 +74,80 @@ public async Task EchoMessage(string message, bool asEmbed = false, string embed
+                embed.Color = Color.Blue;
+                embed.Footer = new EmbedFooterBuilder()
+                {
+                    Text = "Luminate - Your ideas shine bright"
```
Same here the string should be changed

lenajenichen commented on LuminateDiscordBot/Commands.cs:
```diff
@@ -63,10 +74,80 @@ public async Task EchoMessage(string message, bool asEmbed = false, string embed
+            embed.Color = Color.Blue;
+            embed.Footer = new EmbedFooterBuilder()
+            {
+                Text = "Luminate - Your ideas shine bright"
```
Same here the string should be changed

lenajenichen commented on LuminateDiscordBot/Commands.cs:
```diff
@@ -63,10 +74,80 @@ public async Task EchoMessage(string message, bool asEmbed = false, string embed
+                embed.Color = Color.Blue;
+                embed.Footer = new EmbedFooterBuilder()
+                {
+                    Text = $"Luminate - Your ideas shine bright"
```
Same here the string should be changed

lenajenichen commented on LuminateDiscordBot/Commands.cs:
```diff
@@ -77,7 +158,7 @@ public async Task EchoMessageFromFile(IAttachment file, bool asEmbed = false, st
             embed.Footer = new EmbedFooterBuilder()
             {
-                Text = $"Luminate - {Utils.DefaultSloganText}"
+                Text = $"Luminate - Your ideas shine bright"
```
Same here the string should be changed

lenajenichen commented on LuminateDiscordBot/Commands.cs:
```diff
@@ -89,7 +170,7 @@ public async Task EchoMessageFromFile(IAttachment file, bool asEmbed = false, st
                 echo.Footer = new EmbedFooterBuilder()
                 {
-                    Text = $"Luminate - {Utils.DefaultSloganText}"
+                    Text = $"Luminate - Your ideas shine bright"
```
Same here the string should be changed

lenajenichen commented on LuminateDiscordBot/Commands.cs:
```diff
@@ -24,23 +17,41 @@ public async Task ModifyChannelRules([Summary("channel_identifier", "The interna
             embed.Color = Color.Blue;
             embed.Footer = new EmbedFooterBuilder()
             {
-                Text = $"Luminate - {Utils.DefaultSloganText}"
```
Why did we remove the Utils.DefaultSolganText?

lenajenichen commented on LuminateDiscordBot/Commands.cs:
```diff
@@ -24,23 +17,41 @@ public async Task ModifyChannelRules([Summary("channel_identifier", "The interna
+        {
+            DBManager.ModifyRoleConfig(roleIdentifier, targetRole.Id);
+            EmbedBuilder embed = new EmbedBuilder();
+            embed.Title = "Updated!";
```
We can also put this message in a variable since we're using it also more than once.

lenajenichen commented on LuminateDiscordBot/Components.cs:
```diff
@@ -1,15 +1,107 @@
+                embed.Color = Color.Blue;
+                embed.Footer = new EmbedFooterBuilder()
+                {
+                    Text = "Luminate - Your ideas shine bright"
```
Same as in Commands.cs. Not going to continue to mark all of them.

lenajenichen commented on LuminateDiscordBot/DBManager.cs:
```diff
@@ -14,31 +11,174 @@ internal class DBManager
+
+        public static Objects.TicketCategory GetTicketCategoryFromName(string ticketTopic)
+        {
+            Objects.TicketCategory ticketCat = null;
```
Add a "?" so that the code knows it can be null.

lenajenichen commented on LuminateDiscordBot/Models/TicketCreationModalModel.cs:
```diff
@@ -0,0 +1,22 @@
+        [RequiredInput(true)]
+        [InputLabel("State your issue")]
+        [ModalTextInput("problem", Discord.TextInputStyle.Paragraph, "Please enter your issue here", 1, 1000)]
+        public string Reason { get; set; }
```
If Reason can be null, add "?" after "string".
`public string? Reason {get; set;}`

lenajenichen commented on LuminateDiscordBot/Objects/TicketCategory.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public class TicketCategory
+    {
+        public string TicketTopic { get; set; }
```
Same as with the reason. Add "?" if it can be null.

lenajenichen commented on LuminateDiscordBot/Objects/TicketCategory.cs:
```diff
@@ -0,0 +1,24 @@
+    public class TicketCategory
+    {
+        public string TicketTopic { get; set; }
+        public string CategoryAliasList { get; set; }
```
Same here.

lenajenichen commented on LuminateDiscordBot/Objects/TicketCategory.cs:
```diff
@@ -0,0 +1,24 @@
+
+        public List<string> GetCategoryAliases()
+        {
+            return JsonSerializer.Deserialize<List<string>>(Convert.FromBase64String(this.CategoryAliasList));
```
Check if the List is null before returning it. 

`if (string.IsNullOrEmpty(CategoryAliasList))
    {
        // Handle whatever you might need when getting a null list.
        return new List<string>();
    }
    else
    {
        return JsonSerializer.Deserialize<List<string>>(Convert.FromBase64String(CategoryAliasList));
    }`

TheVolpeDev replied:
This function is deprecated since the origin ticket system will not be implemented and is a leftover for utility use, it also never is null

lenajenichen replied:
Problem is it **can** be null, therefore we have to either check if it's null or tell the code that it can't be null.

TheVolpeDev replied:
as stated in DBManager.InitDB, TicketKeywords is a TEXT type with a default value of "[]", in no case will the function not have a parseable string, it will however throw an exception because "[]" is not a valid Base64 string, this will later be refactored and replaced with proper JSON instead (since I have adapted to using Parameterized Queries its secure to do)

lenajenichen reviewed (changes requested):
Please change the stuff if needed, so we can get most of the warnings away.

FabioGaming committed:
Fixed every possible / needed null reference

[ # ] Reverted back to Utils.SloganText
[ # ] Made values nullable / required

Any fields that still throw nullable warnings either give me no option to fix or will just push the problem to the followup function which also is out of my control

FabioGaming committed:
removed warnings

[ # ] Supressed warnings in: TicketCategoryAutoCompleteHandler.cs, InteractionHandler.cs, Program.cs
[ # ] Changed some Object variables to be required
[ # ] DBManager operations now use proper Get functions from SQLiteDataReader
